### PR TITLE
WIP Add sign-ups command + code refactoring

### PR DESCRIPTION
**Pull request by sebgrohn**

### CODE REVIEW

File: src/handlers.utils.js [hunk 1]
```diff
@@ -0,0 +1,56 @@
+const R = require('ramda');
+const SlackTemplate = require('claudia-bot-builder').slackTemplate;
+const { InvalidUserInputError, InvalidCallbackActionError } = require('./handlers.errors');
+
+const validateSendingUser = (userRepository, handler) => async (message, ...extra) => {
```
Comment (sebgrohn) [inline src/handlers.utils.js]: Which variant is most useful, this async one or the sync one below?

Pros (of this one):
* Handler can do other async stuff in parallel.

Cons (of this one):
* More code in handler (need to await promise).
* We need to throw a special exception type for the control flow to work (handled explicitly in `bot.js`).

File: src/handlers.js [hunk 2]
```diff
@@ -51,32 +57,21 @@ const handlers = {
...
-        }
-
-        const { actions, callback_id } = originalRequest;
-        const { value: tournamentId } = R.find(({ name }) => name === 'sign_up')(actions);
-
-        if (!tournamentId) {
-            throw new Error(`Invalid action value(s) for callback: ${callback_id}`);
-        }
-
+    signUpUserCallback: ({ challongeService, userRepository }) => validateCallbackValue('sign_up', validateSendingUser(userRepository, async function (message, userPromise, tournamentId) {
```
Comment (sebgrohn) [inline src/handlers.js]: Could we use `R.pipe` here instead, piping `validateSendingUser` -> `validateCallbackValue` -> actual handler? I assume that we couldn't pass in multiple args like we do now, but rather would extend the message object...

Also, would we be forced to throw errors on validation failure (like for the async `validateSendingUser`?

File: src/handlers.utils.js [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+            if (!user) {
+                throw new InvalidUserInputError(
+                    new SlackTemplate('I don\'t know who you are. :crying_cat_face:')
+                        .replaceOriginal(false)
+                        .get(),
+                );
+            }
+            return user;
+        });
+    return await handler(message, ...extra, userPromise);
```
Comment (sebgrohn) [inline src/handlers.utils.js]: I'm not even sure that this `...extra` trick will work, but the idea was to be able to combine / pipe several `validate*` methods.

**Comment by CurriedFloat** [comment]
> I'd love your comments, @CurriedFloat. [b9f8736](https://github.com/sebgrohn/tournie/commit/b9f87361b0db9660f302d5a486fe9c58c11e3f62) contains the most interesting changes.

The goal is to have a pipeable middleware layer for request validation and transformation?
Similar to how some node.js scaffolds look: `router.get('/', auth.hasRole('admin'), controller.index);`, 
but in a more functional/ramda style?

https://ramdajs.com/docs/#pipeWith looks interesting. because you can stop the pipe without throwing.
Maybe look at using a maybe monad, no pun intended.

<blockquote><div><strong><a href="https://ramdajs.com/docs/#pipeWith">Ramda Documentation</a></strong></div></blockquote>

**Comment by CurriedFloat** [comment]
> Error handling shouldn't be handled by exceptions, but by either monads or promises.

source of quote: https://github.com/haskellcamargo/js-real-world-functional-programming

ramda-fantasy recommends https://github.com/sanctuary-js/sanctuary-either for the either monad.

File: src/handlers.utils.js (hunk 3 above)
Comment (CurriedFloat) [reply src/handlers.utils.js]: How about something like:
```
    const commandHandlers = {
        tournaments: [
          transformReqBody,
          validateUser,
          validateSomeOtherThing, 
          handlers.listOpenTournaments
        ],
        whoami: [handlers.showCurrentUser],
    };
```
Then just pipe/pipeWith the array of functions

**Comment by sebgrohn** [comment]
Okay, I've now read up on monads (again...), and there are a lot of nice concepts that you mention, Isak. Maybe monads seems nice, but how would they work with async results?

**Comment by CurriedFloat** [comment]
> As of 0.26.0, pipeP and composeP got deprecated in favor of ramdajs.com/docs/#pipeWith and ramdajs.com/docs/#composeWith

source of quote: https://stackoverflow.com/questions/33486989/handling-asynchronous-programming-with-ramda

Haven't tried it but, I imagine something like:
```
const pipeWhileSuccessful = R.pipeWith(
  (nextFunc, either) => {
    /* TODO async stuff here!!! */
    either.chain(nextFunc)
  });
const runReq = pipeWhileSuccessful([
  transformReqBody,
  validateUser,
  handleUseCase
])
```

**Comment by sebgrohn** [comment]
I pushed some new code: `chain` (and `concurrent`) in `handlers.utils.js`, with example usages in `handlers.js`. Only using `pipeWith`and promises, though, no Maybe / Either. What d'you think?

**Comment by CurriedFloat** [comment]
Sounds good!

Yeah monadic warping will often affect the type signature of a large part of the call chain...

**Pull request merged**